core-util/SharedPointer.h

```diff
@@ -0,0 +1,275 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2015 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __SHAREDPOINTER_H__
+#define __SHAREDPOINTER_H__
+
+#include "core-util/core-util.h"
+#include "mbed-alloc/ualloc.h"
+
+#include <stdint.h>
+#include <stddef.h>
+
+#ifndef NDEBUG
+#include <stdio.h>
+#define SHAREDPOINTER_DEBUG(...) { printf(__VA_ARGS__); }
+#else
+#define SHAREDPOINTER_DEBUG(...) /* nothing */
+#endif
+
+/*
+    Shared pointer class.
+
+    Similar to std::shared_ptr in C++11.
+
+    Usage: SharedPointer<class> POINTER(new class())
+
+    When POINTER is passed around by value the copy constructor and
+    destructor counts the number of references to the original object.
+    If the counter reaches zero, delete is called on the object pointed to.
+
+    To avoid loops, "weak" references should be used by calling the original
+    pointer directly through POINTER.get().
+*/
+
+template <class T>
+class SharedPointer
+{
```

Comment:
Please follow the mbed OS coding style, see for example https://github.com/ARMmbed/core-util/blob/master/core-util/Array.h